### PR TITLE
ARK implementation TrustWallet

### PR DESCRIPTION
**Pull request by kashtik**
## Description

Added ARK implementation.

## Testing instructions

./tools/tests

WARNING: will generate different private key with BIP44 than their official wallet. Official wallet just hashes the mnemonic instead of using BIP44.

### CODE REVIEW

File: swift/Tests/CoinAddressDerivationTests.swift [hunk 1]
```diff
@@ -157,6 +157,11 @@ class CoinAddressDerivationTests: XCTestCase {
                 case .semux:
                     let expectedResult = "0xfe604170382452f77bc922bc19eb4b53504b09c2"
                     AssetCoinDerivation(coin, expectedResult, derivedAddress, address)
+                case .ark:
+                    let expectedResult = "Ac49m5pu5YpMMNgEbSYeZUEpRMHcSK3DfV"
+                    AssetCoinDerivation(coin, expectedResult, derivedAddress, address)
+                default:
```
Comment (vikmeup) [inline swift/Tests/CoinAddressDerivationTests.swift]: no need to specify `default`, it's done on purpose to make switch exhaustive

File: src/interface/TWARKSigner.cpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    PrivateKey key = PrivateKey(Data(input.private_key().begin(), input.private_key().end()));
+
+    Transaction t = Transaction();
+    
+    t.timestamp = static_cast<uint32_t>(input.timestamp());
+    t.recipientId = input.recipientid();
+    t.amount = static_cast<uint64_t>(input.amount());
+    
+    uint64_t fee = static_cast<uint64_t>(input.fee());
+    if ( fee > 0  ){
```
Comment (vikmeup) [inline src/interface/TWARKSigner.cpp]: what is the purpose of `fee > 0` check?

File: src/ARK/Transaction.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+#include <nlohmann/json.hpp>
+
+
+using json = nlohmann::json;
+
+using namespace TW;
+
+ARK::Transaction::Transaction(){
+    type=0; // transfer
+    fee = 10000000; // default fee
```
Comment (vikmeup) [inline src/ARK/Transaction.cpp]: did you consider to expose these values to protobuf model? so developers have the ability to specify this

File: include/TrustWalletCore/TWBlockchain.h [hunk 4]
```diff
@@ -26,7 +26,8 @@ enum TWBlockchain {
     TWBlockchainAion = 11,
     TWBlockchainCosmos = 12,
     TWBlockchainTheta = 23,
-    TWBlockchainOntology = 14
+    TWBlockchainOntology = 14,
+    TWBlockchainArk
```
Comment (hewigovens) [inline include/TrustWalletCore/TWBlockchain.h]:
```suggestion
    TWBlockchainArk = 15
```

File: src/ARK/Address.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+bool TW::ARK::Address::isValid(const std::string &string ){
+    Data decoded = Base58::bitcoin.decodeCheck(string);
+
+    if ( decoded.size() == 0 ){
+        return false;
+    }
+    
+    // mainnet
+    if (decoded[0]!=0x17){
```
Comment (hewigovens) [inline src/ARK/Address.cpp]: can you use a constant for  `0x17`?

File: src/ARK/Address.cpp [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+TW::ARK::Address::Address(const Data &data){
+    address = Base58::bitcoin.encodeCheck(data);
+}
+
+TW::ARK::Address::Address(const PublicKey &pub){
+    Data hash = Data();
+
+    hash.insert(hash.end(), 0x17);
```
Comment (hewigovens) [inline src/ARK/Address.cpp]: `0x17` is used here again

File: src/ARK/Address.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include "../Data.h"
+#include "../PublicKey.h"
+#include "../HexCoding.h"
+
+
+namespace TW::ARK{
+
+    class Address{
```
Comment (hewigovens) [inline src/ARK/Address.h]: Can you follow other `Address.h` indentation?

File: src/ARK/Signer.cpp [hunk 8]
```diff
@@ -0,0 +1,14 @@
+#include "Signer.h"
+#include "../PrivateKey.h"
+#include "../PublicKey.h"
+#include "../Hash.h"
+#include <TrustWalletCore/TWPublicKeyType.h>
+
+using namespace TW;
+
+void TW::ARK::Signer::sign(PrivateKey &pk, Transaction &t){
```
Comment (hewigovens) [inline src/ARK/Signer.cpp]:
```suggestion
void TW::ARK::Signer::sign(PrivateKey &pk, Transaction &tx){
```

File: src/ARK/Transaction.cpp [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+#include "../HexCoding.h"
+#include <nlohmann/json.hpp>
+
+
+using json = nlohmann::json;
+
+using namespace TW;
+
+ARK::Transaction::Transaction(){
+    type=0; // transfer
```
Comment (hewigovens) [inline src/ARK/Transaction.cpp]:
```suggestion
    type = 0; // transfer
```

File: src/ARK/Transaction.cpp [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    fee = 10000000; // default fee
+}
+
+Data ARK::Transaction::to_bytes(){
```
Comment (hewigovens) [inline src/ARK/Transaction.cpp]:
```suggestion
Data ARK::Transaction::encoded(){
```

File: src/ARK/Transaction.cpp [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    encode64LE(amount, result);
+    encode64LE(fee, result);
+
+    if( signature.size() > 0 ){
+        append(result, signature);
+    }
+    return result;
+}
+
+std::string ARK::Transaction::to_json(){
```
Comment (hewigovens) [inline src/ARK/Transaction.cpp]: maybe `toJson` or `toJSON` ?

**Comment by hewigovens** [comment]
Can you add copyright header for new files?
```bash
// Copyright © 2017-2019 Trust Wallet.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.
```

File: include/TrustWalletCore/TWARKAddress.h [hunk 12]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+#include "TWBase.h"
+#include "TWData.h"
+#include "TWString.h"
+
+TW_EXTERN_C_BEGIN
+
+struct TWPublicKey;
+
+/// Represents an Aion address.
```
Comment (hewigovens) [inline include/TrustWalletCore/TWARKAddress.h]:
```suggestion
/// Represents an ARK address.
```

**Comment by hewigovens** [comment]
can you rebase master and fix the CI? thanks

File: swift/Tests/Blockchains/ARKTests.swift [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    
+    func testSigner() throws {
+        let input = ARKSigningInput.with {
+            $0.amount = 1000000000
+            $0.privateKey = Data(hexString: "d8839c2432bfd0a67ef10a804ba991eabba19f154a3d707917681d45822a5712")!
+            $0.timestamp = 50686854
+            $0.recipientID = "ANBkoGqWeTSiaEVgVzSKZd3jS7UWzv9PSo"
+        }
+        
+        let output = ARKSigner.sign(input: input)
```
[the remark this answers is not in the document]
Comment (kashtik) [reply swift/Tests/Blockchains/ARKTests.swift]: Yeah, I didnt want to risk the order of items in json being different.

File: swift/Tests/CoinAddressDerivationTests.swift (hunk 1 above)
Comment (kashtik) [reply swift/Tests/CoinAddressDerivationTests.swift]: True

File: src/interface/TWARKSigner.cpp (hunk 2 above)
Comment (kashtik) [reply src/interface/TWARKSigner.cpp]: If it gets initialized at 0, it was never set so I would like to use default in this case.

File: src/ARK/Transaction.cpp (hunk 3 above)
Comment (kashtik) [reply src/ARK/Transaction.cpp]: Fee is exposed. Type has to be 0 for transfer. I thought we are limited to transfers only.

File: android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/ark/TestArkTransactionSigning.kt [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+class TestArkTransactionSigning{
+    init {
+        System.loadLibrary("TrustWalletCore")
+    }
+    @Test
+    fun arkTransactionSigning(){
+        val signingInput = ARK.SigningInput.newBuilder()
+            .setAmount(1000000000)
+            .setPrivateKey(ByteString.copyFrom("d8839c2432bfd0a67ef10a804ba991eabba19f154a3d707917681d45822a5712"
```
Comment (eddywm) [inline android/app/src/androidTest/java/com/trustwallet/core/app/blockchains/ark/TestArkTransactionSigning.kt]: There's an extension method for conversion from ByteArray to ByteString.
Have a look in `utils/Extensions.kt`

**Pull request closed without merging**